packages/block-library/src/table-of-contents/edit.js

```diff
@@ -15,127 +15,228 @@ import {
-			// Blocks can be loaded into a *non-post* block editor.
-			// eslint-disable-next-line @wordpress/data-no-store-string-literals
-			postContent: select( 'core/editor' ).getEditedPostContent(),
-		} ),
+	const listBlockExists = useSelect(
+		( select ) => !! select( blocksStore ).getBlockType( 'core/list' ),
 		[]
 	);
 
-	// The page this block would be part of on the front-end. For performance
-	// reasons, this is only calculated when onlyIncludeCurrentPage is true.
-	const pageIndex = useSelect(
-		( select ) => {
-			if ( ! onlyIncludeCurrentPage ) {
-				return null;
-			}
+	const {
+		__unstableMarkNextChangeAsNotPersistent,
+		replaceBlocks,
+	} = useDispatch( blockEditorStore );
 
+	/**
+	 * The latest heading data, or null if the new data deeply equals the saved
+	 * headings attribute.
+	 *
+	 * Since useSelect forces a re-render when its return value is shallowly
+	 * inequal to its prior call, we would be re-rendering this block every time
+	 * the stores change, even if the latest headings were deeply equal to the
+	 * ones saved in the block attributes.
+	 *
+	 * By returning null when they're equal, we reduce that to 2 renders: one
+	 * when there are new latest headings (and so it returns them), and one when
+	 * they haven't changed (so it returns null). As long as the latest heading
+	 * data remains the same, further calls of the useSelect callback will
+	 * continue to return null, thus preventing any forced re-renders.
+	 */
+	const latestHeadings = useSelect(
+		( select ) => {
 			const {
 				getBlockAttributes,
 				getBlockIndex,
 				getBlockName,
 				getBlockOrder,
+				getBlockParents,
+				getClientIdsOfDescendants,
+				__experimentalGetGlobalBlocksByName: getGlobalBlocksByName,
 			} = select( blockEditorStore );
 
-			const blockIndex = getBlockIndex( clientId );
-			const blockOrder = getBlockOrder();
-
-			// Calculate which page the block will appear in on the front-end by
-			// counting how many <!--nextpage--> tags precede it.
-			// Unfortunately, this implementation only accounts for Page Break and
-			// Classic blocks, so if there are any <!--nextpage--> tags in any
-			// other block, they won't be counted. This will result in the table
-			// of contents showing headings from the wrong page if
-			// onlyIncludeCurrentPage === true. Thankfully, this issue only
-			// affects the editor implementation.
-			let page = 1;
-			for ( let i = 0; i < blockIndex; i++ ) {
-				const blockName = getBlockName( blockOrder[ i ] );
+			// Disable reason: blocks can be loaded into a *non-post* block
+			// editor, so to avoid declaring @wordpress/editor as a dependency, we
+			// must access its store by string. When the store is not available,
+			// editorSelectors will be null, and the block's saved markup will
+			// lack permalinks.
+			// eslint-disable-next-line @wordpress/data-no-store-string-literals
+			const editorSelectors = select( 'core/editor' );
```

Comment:
For posterity, this pattern is used in the calendar block, too:

https://github.com/WordPress/gutenberg/blob/6821d2968a9223f7f97a97e6427d7cc09300eb05/packages/block-library/src/calendar/edit.js#L51-L54

And in the `.native.js` versions of the columns, file, and image blocks:

https://github.com/WordPress/gutenberg/blob/6821d2968a9223f7f97a97e6427d7cc09300eb05/packages/block-library/src/columns/edit.native.js#L465

https://github.com/WordPress/gutenberg/blob/6821d2968a9223f7f97a97e6427d7cc09300eb05/packages/block-library/src/file/edit.native.js#L596

https://github.com/WordPress/gutenberg/blob/6821d2968a9223f7f97a97e6427d7cc09300eb05/packages/block-library/src/image/edit.native.js#L854

@ZebulanStanphill I would add a similar comment as in the calendar block and then create a separate issue to invent a canonical way of handling these dependencies. Or perhaps there is one already?

Reply:
Done. I've created #39987 to discuss a proper solution to this.